From 30d7356c61e7894d51f32d1a5d2fd1e42e05e3b2 Mon Sep 17 00:00:00 2001
From: Bernt Budde <bernt.budde@gmail.com>
Date: Sat, 31 Oct 2026 10:30:33 +0200
Subject: [PATCH 90/99] TODO tests: <...> after a builtin list operator is a
 readline, also when the glob holds -DIGIT
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

perl reads <...> after sort, reverse, print or return as a readline or a
glob (toke.c yyl_leftpointy: PL_expect is not XOPERATOR after a builtin
that takes an operand).  PPI reads the < as less-than; with a path such as
</tmp/x-1/y/*.pl> the -1 then becomes a Number and the y/ a
transliteration that swallows the next statement.  PCL §14 (task #4100).
---
 t/ppi_bugs_pcl.t | 15 ++++++++++++++-
 1 file changed, 14 insertions(+), 1 deletion(-)

diff --git a/t/ppi_bugs_pcl.t b/t/ppi_bugs_pcl.t
index 72c29792..6c874028 100644
--- a/t/ppi_bugs_pcl.t
+++ b/t/ppi_bugs_pcl.t
@@ -13,7 +13,7 @@
 use lib 't/lib';
 use PPI::Test::pragmas;
 use utf8;
-use Test::More tests => 295 + ($ENV{AUTHOR_TESTING} ? 1 : 0);
+use Test::More tests => 296 + ($ENV{AUTHOR_TESTING} ? 1 : 0);
 
 use PPI ();
 
@@ -1860,3 +1860,16 @@ TODO: {
     my @c = bracket_classes( 'require *{"\0a"};', '{' );
     is( $c[0], 'PPI::Structure::Block', 'require *{"..."}: the dereference block is a Block' );
 }
+
+# §14, the list-operator half: after a builtin that takes an operand (sort,
+# reverse, print, return) perl expects a term, so `<...>` is a readline or a
+# glob whatever it holds (toke.c yyl_leftpointy), even a path with a -DIGIT
+# in it: `my @b = sort </tmp/x-1/y/*.pl>; print 1;` globs and prints 1.
+# Read as a less-than, the -1 became a Number and the y/ a transliteration
+# that swallowed the next statement.
+TODO: {
+    local $TODO = 'PCL §14: <FH> / <glob> after a list operator or a block is lexed as < … >';
+    my @t = map { short_class($_) } toks('my @b = sort </tmp/x-1/y/*.pl>; print 1;');
+    is( $t[4], 'QuoteLike::Readline=</tmp/x-1/y/*.pl>',
+        'sort </tmp/x-1/y/*.pl>: the glob after sort is one readline' ) or diag "got: @t";
+}
-- 
2.53.0

